drivers/pio/sunxi-pio.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <bitmap.h>
+#include <debug.h>
+#include <dm.h>
+#include <error.h>
+#include <mmio.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <util.h>
+#include <drivers/clock.h>
+#include <drivers/irqchip.h>
+#include <drivers/msgbox.h>
+#include <drivers/msgbox/sunxi-msgbox.h>
+#include <drivers/pio.h>
+#include <drivers/pio/sunxi-pio.h>
+
+#define PIO_CFG_REG(n)        (0x18 & (n)) /* Middle two bits of pin. */
+#define PIO_CFG_REG_OFFSET(n) (0x7 & (n))  /* Last three bits of pin. */
+#define PIO_DATA_REG          0x10
+#define PIO_EINT_CTL_REG      0x210
+#define PIO_EINT_STA_REG      0x214
+#define PL_CFG0_REG_CLR       0x77777777
+#define PL_CFG1_REG_CLR       0x00077777
+
+static int
+sunxi_pio_probe(struct device *dev)
+{
+	int err;
+
+	/* Disable IO for all port L pins. */
+	mmio_write32(dev->regs + PIO_CFG_REG(0), PL_CFG0_REG_CLR);
+	mmio_write32(dev->regs + PIO_CFG_REG(1), PL_CFG1_REG_CLR);
+
+	/* Disable and clear all IRQs. */
+	mmio_setbits32(dev->regs + PIO_EINT_CTL_REG, BITMASK(0, 12));
+	mmio_setbits32(dev->regs + PIO_EINT_STA_REG, BITMASK(0, 12));
```

Comment:
Needs to be done for all ports (you can put the number of ports, or a bitmask of available ports, in `dev->drvdata`)